utils/pattern/src/single.rs

```diff
@@ -0,0 +1,256 @@
+///         .unwrap()
+///         .interpolate_to_string(("hi",)),
+///     "{0} hi",
+/// );
+/// ```
+///
+/// [`Pattern::interpolate()`]: crate::Pattern::interpolate
+#[derive(Debug)]
+pub struct SinglePlaceholder {
+    _not_constructible: core::convert::Infallible,
+}
+
+impl crate::private::Sealed for SinglePlaceholder {}
+
+impl PatternBackend for SinglePlaceholder {
+    type PlaceholderKey = SinglePlaceholderKey;
+    type Store = str;
+    type Iter<'a> = SinglePlaceholderPatternIterator<'a>;
+
+    fn validate_store(store: &Self::Store) -> Result<(), Error> {
+        let placeholder_offset_char = store.chars().next().ok_or(Error::InvalidPattern)?;
+        let initial_offset = char::len_utf8(placeholder_offset_char);
+        let placeholder_offset = placeholder_offset_char as usize;
+        if placeholder_offset > store.len() - initial_offset + 1 {
+            return Err(Error::InvalidPattern);
+        }
+        Ok(())
+    }
+
+    fn iter_items(store: &Self::Store) -> Self::Iter<'_> {
+        let placeholder_offset_char = match store.chars().next() {
+            Some(i) => i,
+            None => {
+                debug_assert!(false);
+                '\0'
+            }
+        };
+        let initial_offset = char::len_utf8(placeholder_offset_char);
+        SinglePlaceholderPatternIterator {
+            store,
+            placeholder_offset: placeholder_offset_char as usize + initial_offset - 1,
+            current_offset: initial_offset,
+        }
+    }
+
+    #[cfg(feature = "alloc")]
+    fn try_from_items<
+        'a,
+        I: Iterator<Item = Result<PatternItemCow<'a, Self::PlaceholderKey>, Error>>,
+    >(
+        items: I,
+    ) -> Result<String, Error> {
+        let mut result = String::new();
+        let mut seen_placeholder = false;
+        for item in items {
+            match item? {
+                PatternItemCow::Literal(s) => result.push_str(&s),
+                PatternItemCow::Placeholder(_) if !seen_placeholder => {
+                    seen_placeholder = true;
+                    let placeholder_offset =
+                        u32::try_from(result.len() + 1).map_err(|_| Error::InvalidPattern)?;
+                    let placeholder_offset_char =
+                        char::try_from(placeholder_offset).map_err(|_| Error::InvalidPattern)?;
+                    result.insert(0, placeholder_offset_char);
+                }
+                PatternItemCow::Placeholder(_) => {
+                    return Err(Error::InvalidPattern);
+                }
+            }
+        }
+        if !seen_placeholder {
+            result.insert(0, '\0');
+        }
+        Ok(result)
+    }
+}
+
+#[doc(hidden)] // TODO(#4467): Should be internal
+#[derive(Debug)]
+pub struct SinglePlaceholderPatternIterator<'a> {
+    store: &'a str,
+    placeholder_offset: usize,
+    current_offset: usize,
+}
+
+impl<'a> Iterator for SinglePlaceholderPatternIterator<'a> {
```

Comment:
issue: if you don't go the array into-iter route, make this implement `ExactSizeIterator`

Reply:
Slightly challenging for two reasons

First, the iterator is returned via

```rust
impl<B, Store> Pattern<B, Store>
where
    B: PatternBackend,
    Store: AsRef<B::Store> + ?Sized,
{
    /// Returns an iterator over the [`PatternItem`]s in this pattern.
    pub fn iter(&self) -> impl Iterator<Item = PatternItem<B::PlaceholderKey>> + '_ {
        B::iter_items(self.store.as_ref())
    }
}
```

I don't want to add more constraints on `impl Iterator` because not all backends may know their length cheaply. I could change this function to instead return the non-generic type `B::Iter` if desired, and then I could implement an additional trait.

Second, returning the length, while a constant-time operation, is not _completely_ free. I either need to store the precomputed length in another field on this iterator type, or always compute it on the fly after re-reading the first code point in the string. I lean toward the second approach.

Reply:
For now I implemented `Iterator::size_hint`

Reply:
I think it wouldn't hurt to implement `ExactSizeIterator`, even if there's currently no way for a user to observe it. The compiler might still be able to use it, as it will monomorphize the `-> impl Iterator` return type to a concrete type at some point.